pkg/builder/builder.go

```diff
@@ -0,0 +1,272 @@
+	buildId := uuid.String()[:8]
+
+	builder := &Builder{
+		id:                buildId,
+		plugin:            nil,
+		project:           p,
+		containerRegistry: cr,
+		gitProviderConfig: gpc,
+
+		daytonaServerConfigFolder:       f.daytonaServerConfigFolder,
+		localContainerRegistryServer:    f.localContainerRegistryServer,
+		basePath:                        f.basePath,
+		loggerFactory:                   f.loggerFactory,
+		defaultProjectImage:             f.defaultProjectImage,
+		defaultProjectUser:              f.defaultProjectUser,
+		defaultProjectPostStartCommands: f.defaultProjectPostStartCommands,
+	}
+
+	return builder
+}
+
+func (b *Builder) GetBuilderPlugin() BuilderPlugin {
+	return b.plugin
+}
+
+func (b *Builder) Prepare() error {
+	hash, err := b.project.GetConfigHash()
+	if err != nil {
+		return err
+	}
+	b.hash = hash
+	projectDir := filepath.Join(b.basePath, hash, "project")
+
+	err = os.RemoveAll(projectDir)
+	if err != nil {
+		return err
+	}
+
+	gitservice := git.Service{
+		ProjectDir:        projectDir,
+		GitConfigFileName: "",
+		//	todo: write to project log
+		LogWriter: os.Stdout,
+	}
+
+	var auth *http.BasicAuth
+	if b.gitProviderConfig != nil {
+		auth = &http.BasicAuth{
+			Username: b.gitProviderConfig.Username,
+			Password: b.gitProviderConfig.Token,
+		}
+	}
+
+	err = gitservice.CloneRepository(&b.project, auth)
+	if err != nil {
+		return err
+	}
+
+	buildConfig := b.project.Build
+
+	if buildConfig != nil && *buildConfig == (workspace.ProjectBuild{}) {
+		//	detect is devcontainer
+		devcontainerPath := ".devcontainer/devcontainer.json"
+		isDevcontainer, err := fileExists(filepath.Join(projectDir, devcontainerPath))
+		if err != nil {
+			return err
+		}
+		if !isDevcontainer {
+			devcontainerPath = ".devcontainer.json"
+			isDevcontainer, err = fileExists(filepath.Join(projectDir, devcontainerPath))
+			if err != nil {
+				return err
+			}
+		}
+		if isDevcontainer {
+			buildConfig.Devcontainer = &workspace.ProjectBuildDevcontainer{
+				DevContainerFilePath: devcontainerPath,
+			}
+			goto initPlugin
+		}
+		//	todo: detect dockerfile
+		//	todo: detect nix
+
+		//	no supported dev config standard found
+		//	set default project image to ensure that project will run anyway
+		b.project.Image = b.defaultProjectImage
+		b.project.User = b.defaultProjectUser
+		b.project.PostStartCommands = b.defaultProjectPostStartCommands
+	}
+
+initPlugin:
+
+	if buildConfig.Devcontainer != nil {
+		b.plugin = &DevcontainerBuilder{
+			DevcontainerBuilderConfig: DevcontainerBuilderConfig{
+				buildId:                      b.id,
+				project:                      b.project,
+				loggerFactory:                b.loggerFactory,
+				localContainerRegistryServer: b.localContainerRegistryServer,
+				projectVolumePath:            filepath.Join(b.basePath, b.hash, "project"),
+			},
+		}
+	}
+
+	return nil
+}
+
+func (b *Builder) LoadBuildResults() (*BuildResult, error) {
+	hash, err := b.project.GetConfigHash()
```

Comment:
Shouldn't we use `b.hash` here?

Also, it might make more sense to move:
```golang
hash, err := b.project.GetConfigHash()
if err != nil {
  return err
}
b.hash = hash
```

This into the `Create` function of the factory.